sky/skylet/providers/aws/config_v2.py

```diff
@@ -0,0 +1,1234 @@
+        profile = _get_instance_profile(instance_profile_name, config)
+        time.sleep(15)  # wait for propagation
+
+    cli_logger.doassert(
+        profile is not None, "Failed to create instance profile."
+    )  # todo: err msg
+    assert profile is not None, "Failed to create instance profile"
+
+    if not profile.roles:
+        role_name = cwh.resolve_iam_role_name(config["provider"], DEFAULT_SKYPILOT_IAM_ROLE)
+        role = _get_role(role_name, config)
+        if role is None:
+            cli_logger.verbose(
+                "Creating new IAM role {} for use as the default instance role.",
+                cf.bold(role_name),
+            )
+            iam = _resource("iam", config)
+            policy_doc = {
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Principal": {"Service": "ec2.amazonaws.com"},
+                        "Action": "sts:AssumeRole",
+                    },
+                ]
+            }
+            attach_policy_arns = cwh.resolve_policy_arns(
+                config["provider"],
+                iam,
+                [
+                    "arn:aws:iam::aws:policy/AmazonEC2FullAccess",
+                    "arn:aws:iam::aws:policy/AmazonS3FullAccess",
+                ],
+            )
+            iam.create_role(
+                RoleName=role_name, AssumeRolePolicyDocument=json.dumps(policy_doc)
+            )
+            
+            role = _get_role(role_name, config)
+            cli_logger.doassert(
+                role is not None, "Failed to create role."
+            )  # todo: err msg
+
+            assert role is not None, "Failed to create role"
+
+            for policy_arn in attach_policy_arns:
+                role.attach_policy(PolicyArn=policy_arn)
+            
+            # SkyPilot: "PassRole" is required by the head node to pass the role to
+            # the workers, so we can access S3 buckets on the workers. "Resource"
+            # is to limit the role to only able to pass itself to the workers.
+            iam_self_read_write_policy_doc = {
+                "Statement": [
+                    {
+                        "Effect": "Allow",
+                        "Action": [
+                            "iam:GetRole",
+                            "iam:PassRole",
+                        ],
+                        "Resource": role.arn,
+                    },
+                    {
+                        "Effect": "Allow",
+                        "Action": "iam:GetInstanceProfile",
+                        "Resource": profile.arn,
+                    }
+                ]
+            }
+            role.Policy("IAMSelfReadWritePolicy").put(
+                PolicyDocument=json.dumps(iam_self_read_write_policy_doc)
+            )
+
+        profile.add_role(RoleName=role.name)
+        time.sleep(15)  # wait for propagation
```

Comment:
Is there a mechanism that checks for propagation? sleep is inherently flaky

Reply:
It was directly from ray's aws `config.py`. Maybe we can try to fix it in the future?
https://github.com/ray-project/ray/blob/5d81cf09b1f6125481c6f2e01f92bb0d23599c9f/python/ray/autoscaler/_private/aws/config.py#L334